### ISSUE TITLE
How to load data of any types and any shapes into textures correctly?

### ISSUE THREAD

**Issue by ifsheldon** [sotrh/learn-wgpu#182]
(Sorry for so many issues and questions from me)
As I am trying to do DVR with webgpu, I need to load 1D and 3D data into 1D and 3D textures. When I want to try out loading 3D data at first, I found there are actually corners I don't know and I have done some trials myself and search on the web but I haven't found anything useful, so I think it will be also great if you can explain these details a bit more in your tutorial.

So, the problem is how we deal with texture data with "any" types (i.e., floats, uints and ints of different bits) and "any" shapes(i.e., 1D, 2D, 3D). We've seen depth textures, normal maps and color textures in your tutorial, and all of them are 2D. 

For example, in my case, I need to load 3D volume data into a 3D texture. The volume has `uint16` as its data format and only one channel. I can, for sure, normalize the data to be in range [0,1] and convert the datatype into like `f32`, but I expect webgpu maybe can somehow do this for me as OpenGL does. And I want to turn on linear interpolations.

In OpenGL, I will write the below, can I do something similar in WebGPU? As you may probably very familiar with the APIs, `GL_R32F` specifies I want the internal format of this texture to be single 32-bit float, which is a property of this texture, and `GL_RED` tells the function that the data has only one channel, and `GL_UNSIGNED_SHORT` tell it that one element of the data is 2-byte. Then, OpenGL will do the normalization for me.
```cpp
glTexParameteri(GL_TEXTURE_3D, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);
glTexParameteri(GL_TEXTURE_3D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
glTexImage3D(GL_TEXTURE_3D, 0, GL_R32F, vol_dim[0], vol_dim[1], vol_dim[2], 0, GL_RED, GL_UNSIGNED_SHORT,
             data_array);
```
I can sort of relate some of the fields in structs in WebGPU to the above settings, but I don't see it work in my trials. I tried something like this, but none works and they gave runtime errors. 
```rust
let texture_size = wgpu::Extent3d {
    width: dimensions.0,
    height: dimensions.1,
    depth_or_array_layers: dimensions.2,
};
let texture = device.create_texture(
    &wgpu::TextureDescriptor {
        size: texture_size,
        mip_level_count: 1, 
        sample_count: 1,
        dimension: wgpu::TextureDimension::D3,
        format: wgpu::TextureFormat::R32Float, // or R16Uint if I use unnormalized uint16 data
        usage: wgpu::TextureUsage::SAMPLED | wgpu::TextureUsage::COPY_DST,
        label: Some("texture3d"),
    }
);

queue.write_texture(
    ImageCopyTexture {
        texture: &texture,
        mip_level: 0,
        origin: Origin3d::ZERO,
    },
    bytemuck::cast_slice(data_f32_slice), // or data_u16_slice
    ImageDataLayout {
        offset: 0,
        bytes_per_row: Some(NonZeroU32::new(4 * dimensions.0).unwrap()), // or 2* dimensions.0 if I use unnormalized uin16 data
        rows_per_image: Some(NonZeroU32::new(dimensions.1).unwrap()),
    },
    texture_size,
);
```
I have a related question here and I think it's important. When we do `queue.write_texture`, how can it determine how to interpret the byte data? That is we seem not to have to specify something like `GL_UNSIGNED_SHORT` in my example, but there are many ways to interpret raw bytes.

Another thing is about samplers. It seems there are some constraints on samplers, but I cannot find it anywhere.
Like below, when I create a bind group for textures, I do something like below. However, in runtime, an fetal error occurs, stating some thing like I have set `sample_type: TextureSampleType::Float { filterable: true }` but the texture format is `TextureFormat::R32Float` or `TextureFormat::R16Uint`.

```rust
let texture_bind_group_layout =
device.create_bind_group_layout(&BindGroupLayoutDescriptor {
    label: Some("Texture binding group layout"),
    entries: &[
        BindGroupLayoutEntry {
            binding: 0,
            visibility: ShaderStage::FRAGMENT,
            ty: BindingType::Texture {
                multisampled: false,
                view_dimension: TextureViewDimension::D3,
                sample_type: TextureSampleType::Float { filterable: true }, // what should I set in my example?
            },
            count: None,
        },
        BindGroupLayoutEntry {
            binding: 1,
            visibility: ShaderStage::FRAGMENT,
            ty: BindingType::Sampler {
                comparison: false,
                filtering: true,
            },
            count: None,
        },
    ],
});
```

So, in summary, where do we need to tweak if we want to load "any"(different shapes and different number types) data into textures? 

Thank you so much!

**Comment by ifsheldon** [comment]
Any suggestions?

**Comment by sotrh** [comment]
So for the `write_texture` part, my guess is that wgpu uses the `TextureFormat` used to create the texture to determine how to interpret the data. It's up to you to ensure that the `byte_per_row` value is correct.

As far as the the sampler goes, the reason that `filterable: true` is required is in `TextureSampleType::Float` is that the sampler  binding is has `filtering: true` as you can see in the code you posted. 

```rust
ty: BindingType::Sampler {
    comparison: false,
    filtering: true,
},
```
I believe if you want to use a different sample type, or not use filtering, you have to set filtering to `false`. The sample type has nothing to do with the TextureFormat (afaik), it tells wgpu how to deliver the texture data in the shader. The `Float` option means that colors will be delivered as floats (r, g, b, and a) between 0.0 and 1.0. I believe this option works with all texture formats.

All be honest, I haven't stayed to far from the beaten path in this particular area. I would consult the devs if you need any super deep insight into these sort of things.

**Comment by ifsheldon** [comment]
Thanks a lot! With your answer and the [discussion in wgpu repo], I can now gather all pieces together!

I suggest we can write more details about texture formats and filterability as a side note in the sections for reference.

**Comment by sotrh** [comment]
This link in particular would likely be helpful to a lot of people: https://gpuweb.github.io/gpuweb/#texture-format-caps. Thanks for hunting down that information!
